README.md
# Wealthsimple-node

Node.js client for Wealthsimple API: https://developers.wealthsimple.com

This is still under development and therefore currently only supports sandbox.

## Installation

```
npm install --save https://github.com/anthfran/wealthsimple-node
```

### Example of Usage

```
const credentials = {
  "client_id": "XXXXXXXXX",
  "client_secret": "XXXXXXXXX",
  "redirect_uri": "https://localhost:3000/auth"
};
const wealthsimple = require('wealthsimple-node').appId(credentials);

wealthsimple.tokenExchange(authCode)
.then(tokens => {
  return Wealthsimple.listAccounts(tokens.access_token)
})
.then(accountResponse => {
  console.log(accountResponse);
  // do something with response
})
.catch(error => {
  // do something with error
});
```


<!-- Generated by documentation.js. Update this documentation by updating the source code. -->

### Table of Contents

-   [tokenExchange][1]
    -   [Parameters][2]
    -   [Examples][3]
-   [tokenRefresh][4]
    -   [Parameters][5]
    -   [Examples][6]
-   [createUser][7]
    -   [Parameters][8]
    -   [Examples][9]
-   [listUsers][10]
    -   [Parameters][11]
    -   [Examples][12]
-   [getUser][13]
    -   [Parameters][14]
    -   [Examples][15]
-   [createPerson][16]
    -   [Parameters][17]
    -   [Examples][18]
-   [listPeople][19]
    -   [Parameters][20]
    -   [Examples][21]
-   [getPerson][22]
    -   [Parameters][23]
    -   [Examples][24]
-   [updatePerson][25]
    -   [Parameters][26]
    -   [Examples][27]
-   [createAccount][28]
    -   [Parameters][29]
    -   [Examples][30]
-   [listAccounts][31]
    -   [Parameters][32]
    -   [Examples][33]
-   [getAccount][34]
    -   [Parameters][35]
    -   [Examples][36]
-   [getAccountTypes][37]
    -   [Parameters][38]
    -   [Examples][39]
-   [getDailyValues][40]
    -   [Parameters][41]
    -   [Examples][42]
-   [listPositions][43]
    -   [Parameters][44]
    -   [Examples][45]
-   [listTransactions][46]
    -   [Parameters][47]
    -   [Examples][48]
-   [getProjection][49]
    -   [Parameters][50]
    -   [Examples][51]
-   [listBankAccounts][52]
    -   [Parameters][53]
    -   [Examples][54]
-   [createDeposit][55]
    -   [Parameters][56]
    -   [Examples][57]
-   [listDeposits][58]
    -   [Parameters][59]
    -   [Examples][60]
-   [getDeposit][61]
    -   [Parameters][62]
    -   [Examples][63]

## tokenExchange

Exchanges an auth code for OAuth2 tokens

### Parameters


-   `code` **[String][64]** Auth string from Wealthsimple redirect

### Examples

```javascript
wealthsimple.tokenExchange(authCode).then(response=>console.log(response));
```

Returns **[Promise][65]** Fetch promise which will resolve containing OAuth2 Tokens

## tokenRefresh

Refreshes OAuth2 tokens

### Parameters


-   `refreshToken` **[String][64]** Refresh Token

### Examples

```javascript
wealthsimple.tokenRefresh(refreshToken).then(response=>console.log(response));
```

Returns **[Promise][65]** Fetch promise which will resolve containing OAuth2 Tokens

## createUser

Create a User
[https://developers.wealthsimple.com/#operation/Create%20User][66]

### Parameters


-   `body` **[Object][67]**

### Examples

```javascript
wealthsimple.createUser(body).then(response=>console.log(response));
```

Returns **[Promise][65]** Fetch promise which will resolve with newly created user

## listUsers

List Users
[https://developers.wealthsimple.com/#operation/List%20Users][68]
This API will return a list of Users scoped by the authorization credentials.

### Parameters


-   `token` **[String][64]** OAuth token for a user
-   `params` **[Object][67]?** See Wealthsimple website for an example of the request parameters

### Examples

```javascript
wealthsimple.listUsers(token).then(response=>console.log(response));
```

```javascript
let params = { limit: 25, offset: 50, created_before: "2017-06-21"};
wealthsimple.listUsers(token, params).then(response=>console.log(response));
```

Returns **[Promise][65]** Fetch promise which will resolve with newly created user

## getUser

Get User
[https://developers.wealthsimple.com/#operation/Get%20User][69]

### Parameters


-   `token` **[String][64]** OAuth token for a user
-   `userId` **[String][64]** Example "user-12398ud"

### Examples

```javascript
wealthsimple.getUser(token, userId).then(response=>console.log(response));
```

Returns **[Promise][65]** Fetch promise which will resolve with user info

## createPerson

Create Person
[https://developers.wealthsimple.com/#operation/Create%20Person][70]

### Parameters


-   `token` **[String][64]** OAuth token for a user
-   `body` **[Object][67]** See Wealthsimple website for an example of the request body

### Examples

```javascript
wealthsimple.createPerson(token, body).then(response=>console.log(response));
```

Returns **[Promise][65]** Fetch promise which will resolve with the created Person

## listPeople

List People
[https://developers.wealthsimple.com/#operation/List%20People][71]
This API will return a list of People scoped by the authorization credentials.

### Parameters


-   `token` **[String][64]** OAuth token for a user
-   `params` **[Object][67]** See Wealthsimple website for an example of the request parameters

### Examples

```javascript
wealthsimple.createPerson(token, body).then(response=>console.log(response));
```

Returns **[Promise][65]** Fetch promise which will resolve with the list of people

## getPerson

Get Person
[https://developers.wealthsimple.com/#operation/Get%20Person][72]
Get a Person entity if you know the person_id and the current credentials have access to the Person.

### Parameters


-   `token` **[String][64]** OAuth token for a user
-   `personId` **[String][64]** Example "person-12398ud"

### Examples

```javascript
wealthsimple.getPerson(token, "person-12398ud").then(response=>console.log(response));
```

Returns **[Promise][65]** Fetch promise which will resolve with the Person details

## updatePerson

Update Person
[https://developers.wealthsimple.com/#operation/Update%20Person][73]
You can add/remove information to the Person entity as the information becomes available using this API. To remove a previously set attribute, set the value to null. Attributes that are not mentioned in the request payload will leave the attribute unchanged in the Person entity.

### Parameters


-   `token` **[String][64]** OAuth token for a user
-   `personId` **[String][64]** Example "person-12398ud"
-   `body` **[Object][67]** See Wealthsimple website for an example of the body

### Examples

```javascript
wealthsimple.updatePerson(token, "person-12398ud", body).then(response=>console.log(response));
```

Returns **[Promise][65]** Fetch promise which will resolve with the updated Person

## createAccount

Create Account
[https://developers.wealthsimple.com/#operation/Create%20Account][74]
You can add/remove information to the Person entity as the information becomes available using this API. To remove a previously set attribute, set the value to null. Attributes that are not mentioned in the request payload will leave the attribute unchanged in the Person entity.

### Parameters


-   `token` **[String][64]** OAuth token for a user
-   `personId` **[String][64]** Example "person-12398ud"
-   `body` **[Object][67]** See Wealthsimple website for an example of the body

### Examples

```javascript
wealthsimple.createAccount(token, body).then(response=>console.log(response));
```

Returns **[Promise][65]** Fetch promise which will resolve with the created Account

## listAccounts

List Accounts
[https://developers.wealthsimple.com/#operation/List%20Accounts][75]

### Parameters


-   `token` **[String][64]** OAuth token for a user
-   `params` **[Object][67]?** Optional filter params, See Wealthsimple website for an example of the request parameters

### Examples

```javascript
wealthsimple.listAccounts(token).then(response=>console.log(response));
```

```javascript
wealthsimple.listAccounts(token,params).then(response=>console.log(response));
```

Returns **[Promise][65]** Fetch promise which will resolve with the list of accounts

## getAccount

Get Account
[https://developers.wealthsimple.com/#operation/Get%20Account][76]

### Parameters


-   `token` **[String][64]** OAuth token for a user
-   `accountId` **[String][64]** Account ID String

### Examples

```javascript
wealthsimple.listAccounts(token).then(response=>console.log(response));
```

```javascript
wealthsimple.getAccount(token,accountId).then(response=>console.log(response));
```

Returns **[Promise][65]** Fetch promise which will resolve with the account details

## getAccountTypes

Get Account Types
[https://developers.wealthsimple.com/#operation/Get%20Account%20Types][77]
Returns openable account types. If a client_id is provided it will scope the types to the client in question, otherwise it will default to the requestor

### Parameters


-   `token` **[String][64]** OAuth token for a user
-   `params` **[Object][67]?** Optional filter params, See Wealthsimple website for an example of the request parameters

### Examples

```javascript
wealthsimple.getAccountTypes(token).then(response=>console.log(response));
```

```javascript
wealthsimple.getAccountTypes(token,params).then(response=>console.log(response));
```

Returns **[Promise][65]** Fetch promise which will resolve with the account details

## getDailyValues

Get Daily Values
[https://developers.wealthsimple.com/#operation/List%20Daily%20Values][78]
Returns historical daily values for a given account. This API will only return a maximum of 365 days worth of daily values from a given start date. By default, it will return historical values for the last 30-days. The start date must occur before the end date if provided. If the difference between the start date and the end date exceeds 365 days, an error will be thrown. The number of Daily Values can be potentially prohibitively large, the results are paginated.

### Parameters


-   `token` **[String][64]** OAuth token for a user
-   `params` **[Object][67]?** Optional filter params, See Wealthsimple website for an example of the request parameters
    -   `params.accound_id` **[Object][67]** Required account_id param

### Examples

```javascript
wealthsimple.getDailyValues(token,{ params.accound_id: "rrsp-r3e9c1w" }).then(response=>console.log(response));
```

Returns **[Promise][65]** Fetch promise which will resolve with the account daily values

## listPositions

List Positions
[https://developers.wealthsimple.com/#tag/Positions][79]
Returns positions for a given account. This API will also allow you to retrieve historical Positions held on a given date.

### Parameters


-   `token` **[String][64]** OAuth token for a user
-   `params` **[Object][67]?** Optional filter params, See Wealthsimple website for an example of the request parameters
    -   `params.accound_id` **[Object][67]** Required account_id param

### Examples

```javascript
wealthsimple.listPositions(token,{ params.accound_id: "rrsp-r3e9c1w" }).then(response=>console.log(response));
```

Returns **[Promise][65]** Fetch promise which will resolve with the account positions

## listTransactions

List Transactions
[https://developers.wealthsimple.com/#operation/List%20Transactions][80]
Lists all Transactions. The number of Transactions can be potentially prohibitively large, the results are paginated. By default, the API will return the 250 latest transactions in the last 30 days.

### Parameters


-   `token` **[String][64]** OAuth token for a user
-   `params` **[Object][67]?** Optional filter params, See Wealthsimple website for an example of the request parameters
    -   `params.accound_id` **[Object][67]** Required account_id param

### Examples

```javascript
wealthsimple.listTransactions(token,{ params.accound_id: "rrsp-r3e9c1w" }).then(response=>console.log(response));
```

Returns **[Promise][65]** Fetch promise which will resolve with the account transactions

## getProjection

Get Projection
[https://developers.wealthsimple.com/#operation/Get%20Projection][81]
Retrieves a projections of returns for an account based on deposits and frequency.

### Parameters


-   `token` **[String][64]** OAuth token for a user
-   `params` **[Object][67]** Projection params
    -   `params.accound_id` **[Object][67]** Required account_id param
    -   `params.amount` **[Object][67]** Required deposit amount
    -   `params.frequency` **[Object][67]** Required deposit frequency
    -   `params.start_date` **[Object][67]** Required deposit start date

### Examples

```javascript
wealthsimple.getProjection(token, params).then(response=>console.log(response));
```

Returns **[Promise][65]** Fetch promise which will resolve with the projection

## listBankAccounts

List Bank Accounts
[https://developers.wealthsimple.com/#operation/List%20Bank%20Accounts][82]

### Parameters


-   `token` **[String][64]** OAuth token for a user
-   `params` **[Object][67]?** See website for optional query params

### Examples

```javascript
wealthsimple.listBankAccounts(token, params).then(response=>console.log(response));
```

Returns **[Promise][65]** Fetch promise which will resolve with list of bank accounts

## createDeposit

Create Deposit
[https://developers.wealthsimple.com/#operation/Create%20Deposit][83]
Initiates an electronic funds transfer to deposit funds to an Account from a Bank Account

### Parameters


-   `token` **[String][64]** OAuth token for a user
-   `body` **[Object][67]** Required deposit details
    -   `body.bank_account_id` **[Object][67]** The unique id of the Bank Account
    -   `body.account_id` **[Object][67]** The unique id of the Account
    -   `body.amount` **[Object][67]** Dollar amount
    -   `body.currency` **[Object][67]** Currency

### Examples

```javascript
wealthsimple.createDeposit(token, body).then(response=>console.log(response));
```

Returns **[Promise][65]** Fetch promise which will resolve with deposit info

## listDeposits

List Deposits
[https://developers.wealthsimple.com/#operation/List%20Deposits][84]

### Parameters


-   `token` **[String][64]** OAuth token for a user
-   `params` **[Object][67]?** See website for optional query params

### Examples

```javascript
wealthsimple.listDeposits(token, body).then(response=>console.log(response));
```

Returns **[Promise][65]** Fetch promise which will resolve with deposit list

## getDeposit

Get Deposit
[https://developers.wealthsimple.com/#operation/List%20Deposits][84]

### Parameters


-   `token` **[String][64]** OAuth token for a user
-   `fundsTransferId` **[String][64]** funds_transfer_id

### Examples

```javascript
let fundsTransferId = "funds_transfer_id-r3e9c1w";
wealthsimple.getDeposit(token, fundsTransferId).then(response=>console.log(response));
```

Returns **[Promise][65]** Fetch promise which will resolve with a deposit entity.

[1]: #tokenexchange

[2]: #parameters

[3]: #examples

[4]: #tokenrefresh

[5]: #parameters-1

[6]: #examples-1

[7]: #createuser

[8]: #parameters-2

[9]: #examples-2

[10]: #listusers

[11]: #parameters-3

[12]: #examples-3

[13]: #getuser

[14]: #parameters-4

[15]: #examples-4

[16]: #createperson

[17]: #parameters-5

[18]: #examples-5

[19]: #listpeople

[20]: #parameters-6

[21]: #examples-6

[22]: #getperson

[23]: #parameters-7

[24]: #examples-7

[25]: #updateperson

[26]: #parameters-8

[27]: #examples-8

[28]: #createaccount

[29]: #parameters-9

[30]: #examples-9

[31]: #listaccounts

[32]: #parameters-10

[33]: #examples-10

[34]: #getaccount

[35]: #parameters-11

[36]: #examples-11

[37]: #getaccounttypes

[38]: #parameters-12

[39]: #examples-12

[40]: #getdailyvalues

[41]: #parameters-13

[42]: #examples-13

[43]: #listpositions

[44]: #parameters-14

[45]: #examples-14

[46]: #listtransactions

[47]: #parameters-15

[48]: #examples-15

[49]: #getprojection

[50]: #parameters-16

[51]: #examples-16

[52]: #listbankaccounts

[53]: #parameters-17

[54]: #examples-17

[55]: #createdeposit

[56]: #parameters-18

[57]: #examples-18

[58]: #listdeposits

[59]: #parameters-19

[60]: #examples-19

[61]: #getdeposit

[62]: #parameters-20

[63]: #examples-20

[64]: https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/String

[65]: https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Promise

[66]: https://developers.wealthsimple.com/#operation/Create%20User

[67]: https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Object

[68]: https://developers.wealthsimple.com/#operation/List%20Users

[69]: https://developers.wealthsimple.com/#operation/Get%20User

[70]: https://developers.wealthsimple.com/#operation/Create%20Person

[71]: https://developers.wealthsimple.com/#operation/List%20People

[72]: https://developers.wealthsimple.com/#operation/Get%20Person

[73]: https://developers.wealthsimple.com/#operation/Update%20Person

[74]: https://developers.wealthsimple.com/#operation/Create%20Account

[75]: https://developers.wealthsimple.com/#operation/List%20Accounts

[76]: https://developers.wealthsimple.com/#operation/Get%20Account

[77]: https://developers.wealthsimple.com/#operation/Get%20Account%20Types

[78]: https://developers.wealthsimple.com/#operation/List%20Daily%20Values

[79]: https://developers.wealthsimple.com/#tag/Positions

[80]: https://developers.wealthsimple.com/#operation/List%20Transactions

[81]: https://developers.wealthsimple.com/#operation/Get%20Projection

[82]: https://developers.wealthsimple.com/#operation/List%20Bank%20Accounts

[83]: https://developers.wealthsimple.com/#operation/Create%20Deposit

[84]: https://developers.wealthsimple.com/#operation/List%20Deposits


CHANGELOG.md
# Changelog

## v0.1.0 (04 June 2018)

* First version Node client for Wealthsimple API


index.js
const fetch = require('node-fetch');
const queryString = require('query-string');

const sandboxUrl = "https://api.sandbox.wealthsimple.com/v1";

const request = (api, token, params, body) => {
  const postParams = queryString.stringify(params);
  //debug logging
  console.log("url:", sandboxUrl + api.url + "?" + postParams);
  if (body) console.log("body:", JSON.stringify(body));

  return fetch(sandboxUrl + api.url + "?" + postParams, {
    method: api.method,
    headers: {
      'Accept': 'application/json',
      'Content-Type': 'application/json',
      'Authorization': 'Bearer ' + token
    },
    body: JSON.stringify(body)
  })
  .then(response => {
    return response.json()
  });
}

// const healthCheck = request({}, {url:'/healthcheck', method:"GET"});

/**
 * Exchanges an auth code for OAuth2 tokens
 * @param {String} code - Auth string from Wealthsimple redirect
 * @returns {Promise} Fetch promise which will resolve containing OAuth2 Tokens
 * @example
 * wealthsimple.tokenExchange(authCode).then(response=>console.log(response));
 */
const tokenExchange = (appCredentials) => {
  return (code) => {
    let postParams = Object.assign({}, appCredentials, {grant_type: "authorization_code", code: code});
    return request({url:'/oauth/token', method:"POST"}, "", postParams);
  }
}

/**
 * Refreshes OAuth2 tokens
 * @param {String} refreshToken - Refresh Token
 * @returns {Promise} Fetch promise which will resolve containing OAuth2 Tokens
 * @example
 * wealthsimple.tokenRefresh(refreshToken).then(response=>console.log(response));
 */
const tokenRefresh = (appCredentials) => {
  return (refreshToken) => {
    let postParams = Object.assign({}, appCredentials, {grant_type: "refresh_token", refresh_token: refreshToken});
    return request({url:'/oauth/token', method:"POST"}, "", postParams);
  }
}

/**
 * Create a User
 * https://developers.wealthsimple.com/#operation/Create%20User
 *
 * @param {Object} body
 * @returns {Promise} Fetch promise which will resolve with newly created user
 * @example
 * wealthsimple.createUser(body).then(response=>console.log(response));
 */
const createUser = (appCredentials) => {
  return (body) => request({url:'/users', method:"POST"}, "", appCredentials, body);
}

/**
 * List Users
 * https://developers.wealthsimple.com/#operation/List%20Users
 * This API will return a list of Users scoped by the authorization credentials.
 *
 * @param {String} token - OAuth token for a user
 * @param {Object} [params] - See Wealthsimple website for an example of the request parameters
 * @returns {Promise} Fetch promise which will resolve with newly created user
 * @example
 * wealthsimple.listUsers(token).then(response=>console.log(response));
 *
 * @example
 * let params = { limit: 25, offset: 50, created_before: "2017-06-21"};
 * wealthsimple.listUsers(token, params).then(response=>console.log(response));
 */
const listUsers = () => {
  return (token, params) => request({url:'/users', method:"GET"}, token, params);
}

/**
 * Get User
 * https://developers.wealthsimple.com/#operation/Get%20User
 *
 * @param {String} token - OAuth token for a user
 * @param {String} userId - Example "user-12398ud"
 *
 * @returns {Promise} Fetch promise which will resolve with user info
 * @example
 * wealthsimple.getUser(token, userId).then(response=>console.log(response));
 */
const getUser = () => {
  return (token, userId) => request({url:'/users/' + userId, method:"GET"}, token);
}

/**
 * Create Person
 * https://developers.wealthsimple.com/#operation/Create%20Person
 *
 * @param {String} token - OAuth token for a user
 * @param {Object} body - See Wealthsimple website for an example of the request body
 * @returns {Promise} Fetch promise which will resolve with the created Person
 * @example
 * wealthsimple.createPerson(token, body).then(response=>console.log(response));
 */
const createPerson = () => {
  return (token, body) => request({url:'/people', method:"POST"}, token, {}, body);
}
/**
 * List People
 * https://developers.wealthsimple.com/#operation/List%20People
 * This API will return a list of People scoped by the authorization credentials.
 *
 * @param {String} token - OAuth token for a user
 * @param {Object} params - See Wealthsimple website for an example of the request parameters
 * @returns {Promise} Fetch promise which will resolve with the list of people
 * @example
 * wealthsimple.createPerson(token, body).then(response=>console.log(response));
 */
const listPeople = () => {
  return (token, params) => request({url:'/people', method:"GET"}, token, params);
}

/**
 * Get Person
 * https://developers.wealthsimple.com/#operation/Get%20Person
 * Get a Person entity if you know the person_id and the current credentials have access to the Person.
 *
 * @param {String} token - OAuth token for a user
 * @param {String} personId - Example "person-12398ud"
 * @returns {Promise} Fetch promise which will resolve with the Person details
 * @example
 * wealthsimple.getPerson(token, "person-12398ud").then(response=>console.log(response));
 */
const getPerson = () => {
  return (token, personId) => request({url:'/people/' + personId, method:"POST"}, token);
}
/**
 * Update Person
 * https://developers.wealthsimple.com/#operation/Update%20Person
 * You can add/remove information to the Person entity as the information becomes available using this API. To remove a previously set attribute, set the value to null. Attributes that are not mentioned in the request payload will leave the attribute unchanged in the Person entity.
 *
 * @param {String} token - OAuth token for a user
 * @param {String} personId - Example "person-12398ud"
 * @param {Object} body - See Wealthsimple website for an example of the body
 * @returns {Promise} Fetch promise which will resolve with the updated Person
 * @example
 * wealthsimple.updatePerson(token, "person-12398ud", body).then(response=>console.log(response));
 */
const updatePerson = () => {
  return (token, personId, body) => request({url:'/people/' + personId, method:"PATCH"}, token, {}, body);
}

/**
 * Create Account
 * https://developers.wealthsimple.com/#operation/Create%20Account
 * You can add/remove information to the Person entity as the information becomes available using this API. To remove a previously set attribute, set the value to null. Attributes that are not mentioned in the request payload will leave the attribute unchanged in the Person entity.
 *
 * @param {String} token - OAuth token for a user
 * @param {String} personId - Example "person-12398ud"
 * @param {Object} body - See Wealthsimple website for an example of the body
 * @returns {Promise} Fetch promise which will resolve with the created Account
 * @example
 * wealthsimple.createAccount(token, body).then(response=>console.log(response));
 */
const createAccount = () => {
  return (token, body) => request({url:'/accounts', method:"POST"}, token, {}, body);
}

/**
 * List Accounts
 * https://developers.wealthsimple.com/#operation/List%20Accounts
 *
 * @param {String} token - OAuth token for a user
 * @param {Object} [params] - Optional filter params, See Wealthsimple website for an example of the request parameters
 * @returns {Promise} Fetch promise which will resolve with the list of accounts
 * @example
 * wealthsimple.listAccounts(token).then(response=>console.log(response));
 * @example
 * wealthsimple.listAccounts(token,params).then(response=>console.log(response));
 */
const listAccounts = () => {
  return (token, params) => request({url:'/accounts', method:"GET"}, token, params);
}

/**
 * Get Account
 * https://developers.wealthsimple.com/#operation/Get%20Account
 *
 * @param {String} token - OAuth token for a user
 * @param {String} accountId - Account ID String
 * @returns {Promise} Fetch promise which will resolve with the account details
 * @example
 * wealthsimple.listAccounts(token).then(response=>console.log(response));
 * @example
 * wealthsimple.getAccount(token,accountId).then(response=>console.log(response));
 */
const getAccount = () => {
  return (token, accountId) => request({url:'/accounts/' + accountId, method:"GET"}, token);
}

/**
 * Get Account Types
 * https://developers.wealthsimple.com/#operation/Get%20Account%20Types
 * Returns openable account types. If a client_id is provided it will scope the types to the client in question, otherwise it will default to the requestor
 *
 * @param {String} token - OAuth token for a user
 * @param {Object} [params] - Optional filter params, See Wealthsimple website for an example of the request parameters
 * @returns {Promise} Fetch promise which will resolve with the account details
 * @example
 * wealthsimple.getAccountTypes(token).then(response=>console.log(response));
 * @example
 * wealthsimple.getAccountTypes(token,params).then(response=>console.log(response));
 */
const getAccountTypes = () => {
  return (token, params) => request({url:'/accounts/account_types', method:"GET"}, token, params);
}

/**
 * Get Daily Values
 * https://developers.wealthsimple.com/#operation/List%20Daily%20Values
 * Returns historical daily values for a given account. This API will only return a maximum of 365 days worth of daily values from a given start date. By default, it will return historical values for the last 30-days. The start date must occur before the end date if provided. If the difference between the start date and the end date exceeds 365 days, an error will be thrown. The number of Daily Values can be potentially prohibitively large, the results are paginated.
 *
 * @param {String} token - OAuth token for a user
 * @param {Object} [params] - Optional filter params, See Wealthsimple website for an example of the request parameters
 * @param {Object} params.accound_id - Required account_id param
 *
 * @returns {Promise} Fetch promise which will resolve with the account daily values
 * @example
 * wealthsimple.getDailyValues(token,{ params.accound_id: "rrsp-r3e9c1w" }).then(response=>console.log(response));
 */
const getDailyValues = () => {
  return (token, params) => request({url:'/daily_values/', method:"GET"}, token, params);
}

/**
 * List Positions
 * https://developers.wealthsimple.com/#tag/Positions
 * Returns positions for a given account. This API will also allow you to retrieve historical Positions held on a given date.
 *
 * @param {String} token - OAuth token for a user
 * @param {Object} [params] - Optional filter params, See Wealthsimple website for an example of the request parameters
 * @param {Object} params.accound_id - Required account_id param
 *
 * @returns {Promise} Fetch promise which will resolve with the account positions
 * @example
 * wealthsimple.listPositions(token,{ params.accound_id: "rrsp-r3e9c1w" }).then(response=>console.log(response));
 */
const listPositions = () => {
  return (token, params) => request({url:'/positions/', method:"GET"}, token, params);
}

/**
 * List Transactions
 * https://developers.wealthsimple.com/#operation/List%20Transactions
 * Lists all Transactions. The number of Transactions can be potentially prohibitively large, the results are paginated. By default, the API will return the 250 latest transactions in the last 30 days.
 *
 * @param {String} token - OAuth token for a user
 * @param {Object} [params] - Optional filter params, See Wealthsimple website for an example of the request parameters
 * @param {Object} params.accound_id - Required account_id param
 *
 * @returns {Promise} Fetch promise which will resolve with the account transactions
 * @example
 * wealthsimple.listTransactions(token,{ params.accound_id: "rrsp-r3e9c1w" }).then(response=>console.log(response));
 */
const listTransactions = () => {
  return (token, params) => request({url:'/transactions/', method:"GET"}, token, params);
}

/**
 * Get Projection
 * https://developers.wealthsimple.com/#operation/Get%20Projection
 * Retrieves a projections of returns for an account based on deposits and frequency.
 *
 * @param {String} token - OAuth token for a user
 * @param {Object} params - Projection params
 * @param {Object} params.accound_id - Required account_id param
 * @param {Object} params.amount - Required deposit amount
 * @param {Object} params.frequency - Required deposit frequency
 * @param {Object} params.start_date - Required deposit start date
 *
 * @returns {Promise} Fetch promise which will resolve with the projection
 * @example
 * wealthsimple.getProjection(token, params).then(response=>console.log(response));
 */
const getProjection = () => {
  return (token, params) => request({url:'/projections', method:"GET"}, token, params);
}


/**
 * List Bank Accounts
 * https://developers.wealthsimple.com/#operation/List%20Bank%20Accounts
 *
 * @param {String} token - OAuth token for a user
 * @param {Object} [params] - See website for optional query params
 * @returns {Promise} Fetch promise which will resolve with list of bank accounts
 * @example
 * wealthsimple.listBankAccounts(token, params).then(response=>console.log(response));
 */
const listBankAccounts = () => {
  return (token, params) => request({url:'/bank_accounts', method:"GET"}, token, params);
}

/**
 * Create Deposit
 * https://developers.wealthsimple.com/#operation/Create%20Deposit
 * Initiates an electronic funds transfer to deposit funds to an Account from a Bank Account
 *
 * @param {String} token - OAuth token for a user
 * @param {Object} body - Required deposit details
 * @param {Object} body.bank_account_id - The unique id of the Bank Account
 * @param {Object} body.account_id - The unique id of the Account
 * @param {Object} body.amount - Dollar amount
 * @param {Object} body.currency - Currency
 * @returns {Promise} Fetch promise which will resolve with deposit info
 * @example
 * wealthsimple.createDeposit(token, body).then(response=>console.log(response));
 */
const createDeposit = () => {
  return (token, body) => request({url:'/deposits', method:"POST"}, token, {}, body);
}

/**
 * List Deposits
 * https://developers.wealthsimple.com/#operation/List%20Deposits
 *
 * @param {String} token - OAuth token for a user
 * @param {Object} [params] - See website for optional query params
 * @returns {Promise} Fetch promise which will resolve with deposit list
 * @example
 * wealthsimple.listDeposits(token, body).then(response=>console.log(response));
 */
const listDeposits = () => {
  return (token, params) => request({url:'/deposits', method:"GET"}, token, params);
}

/**
 * Get Deposit
 * https://developers.wealthsimple.com/#operation/List%20Deposits
 *
 * @param {String} token - OAuth token for a user
 * @param {String} fundsTransferId - funds_transfer_id
 * @returns {Promise} Fetch promise which will resolve with a deposit entity.
 * @example
 * let fundsTransferId = "funds_transfer_id-r3e9c1w";
 * wealthsimple.getDeposit(token, fundsTransferId).then(response=>console.log(response));
 */
const getDeposit = () => {
  return (token, fundsTransferId) => request({url:'/deposits/' + depositId, method:"GET"}, token);
}


module.exports = {
  appId(appCredentials) {
    if (typeof appCredentials.client_id === "string"
    && typeof appCredentials.client_secret === "string"
    && typeof appCredentials.redirect_uri === "string" ) {
      return {
        // healthCheck: healthCheck,
        /* AUTH */
        tokenExchange: tokenExchange(appCredentials),
        tokenRefresh: tokenRefresh(appCredentials),
        /* USERS */
        createUser: createUser(appCredentials),
        listUsers: listUsers(),
        getUser: getUser(),
        /* PEOPLE */
        listPeople: listPeople(),
        /* ACCOUNTS */
        listAccounts: listAccounts(),
        getAccount: getAccount(),
        getAccountTypes: getAccountTypes(),
        /* DAILY VALUES */
        getDailyValues: getDailyValues(),
        /* POSITIONS */
        listPositions: listPositions(),
        /* TRANSACTIONS */
        listTransactions: listTransactions(),
        /* PROJECTIONS */
        getProjection: getProjection(),
        /* BANK ACCOUNTS */
        listBankAccounts: listBankAccounts(),
        /* DEPOSITS */
        createDeposit: createDeposit(),
        listDeposits: listDeposits(),
        getDeposit: getDeposit(),
      };
    } else {
      console.log("Credentials:", appCredentials);
      throw new Error("Invalid Wealthsimple app credentials");
    }

  }
}
